pantheon/src/main/java/tech/pegasys/pantheon/chainimport/ChainImporter.java

```diff
@@ -0,0 +1,251 @@
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.fasterxml.jackson.databind.MapperFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class ChainImporter<C> {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final ObjectMapper mapper;
+  private final PantheonController<C> controller;
+
+  public ChainImporter(final PantheonController<C> controller) {
+    this.controller = controller;
+    mapper = new ObjectMapper();
+    // Jdk8Module allows us to easily parse {@code Optional} values from json
+    mapper.registerModule(new Jdk8Module());
+    // Ignore casing of properties
+    mapper.configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true);
+  }
+
+  public void importChain(final String chainJson) throws IOException {
+    checkDatabaseIsEmpty();
+
+    ChainData chainData = mapper.readValue(chainJson, ChainData.class);
+
+    final List<Block> importedBlocks = new ArrayList<>();
+    for (final BlockData blockData : chainData.getBlocks()) {
+      final BlockHeader parentHeader = getParentHeader(blockData, importedBlocks);
+      final Block importedBlock = processBlockData(blockData, parentHeader);
+      importedBlocks.add(importedBlock);
+    }
+
+    // Check that imported blocks form the canonical chain, if not print a warning
+    this.checkImportedBlocksAreOnCanonicalChain(importedBlocks);
+  }
+
+  private Block processBlockData(final BlockData blockData, final BlockHeader parentHeader) {
+    LOG.info(
+        "Preparing to import block at height {} (parent: {})",
+        parentHeader.getNumber() + 1L,
+        parentHeader.getHash());
+
+    final WorldState worldState =
+        controller
+            .getProtocolContext()
+            .getWorldStateArchive()
+            .get(parentHeader.getStateRoot())
+            .get();
+    final List<Transaction> transactions =
+        blockData.streamTransactions(worldState).collect(Collectors.toList());
+
+    final Block block = createBlock(blockData, parentHeader, transactions);
+    assertAllTransactionsIncluded(block, transactions);
+    importBlock(block);
+
+    return block;
+  }
+
+  private Block createBlock(
+      final BlockData blockData,
+      final BlockHeader parentHeader,
+      final List<Transaction> transactions) {
+    final MiningCoordinator miner = controller.getMiningCoordinator();
+    final GenesisConfigOptions genesisConfigOptions = controller.getGenesisConfigOptions();
+    setOptionalFields(miner, blockData, genesisConfigOptions);
+
+    Optional<Block> maybeBlock =
+        miner.createBlock(parentHeader, transactions, Collections.emptyList());
+
+    if (maybeBlock.isEmpty()) {
+      // Some MiningCoordinator's (specific to consensus type) do not support block-level imports
```

Comment:
```suggestion
      // Some MiningCoordinators (specific to consensus type) do not support block-level imports
```

Reply:
Since `MiningCoordinator` is a class not a word, I usually pluralize with `'s` 🤷‍♀